### PR TITLE
server: fix data race in TestIssue33699

### LINKED ISSUE #33983: data race in server.TestIssue33699
cd server
go test -c -race .
./server.test -test.run TestIssue33699

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #33983

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

Check no data race run this command:

```
make ut X='run server  TestIssue33699 --race'
```

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by hawkingrei — APPROVED** [review]

File: server/driver_tidb.go [hunk 1]
```diff
@@ -196,11 +196,10 @@ func (qd *TiDBDriver) OpenCtx(connID uint64, capability uint32, collation uint8,
 	}
 	se.SetClientCapability(capability)
 	se.SetConnectionID(connID)
-	tc := &TiDBContext{
-		Session:   se,
-		currentDB: dbname,
-		stmts:     make(map[int]*TiDBStatement),
-	}
+	tc := new(TiDBContext)
```
Comment (ngaut) [inline server/driver_tidb.go:199]: Please use the style on left side.

Comment (tiancaiamao) [reply server/driver_tidb.go]: Done.
I don't meant to change here, just to get the exact line to figure out how it races.

File: server/conn.go [hunk 2]
```diff
@@ -202,6 +207,18 @@ type clientConn struct {
 	}
 }
 
+func (cc *clientConn) getCtx() *TiDBContext {
+	cc.ctx.RLock()
+	defer cc.ctx.RUnlock()
+	return cc.ctx.TiDBContext
+}
+
+func (cc *clientConn) setCtx(ctx *TiDBContext) {
```
Comment (bb7133) [inline server/conn.go:216]: Please change it to:

```
	cc.ctx.Lock()
	cc.ctx.TiDBContext = ctx
    cc.ctx.Unlock()
```

Comment (tiancaiamao) [reply server/conn.go:216]: Done

**Review by bb7133 — APPROVED** [review]
LGTM

**Pull request merged**